enterprise/internal/database/perms_store_test.go

```diff
@@ -668,6 +672,331 @@ func testPermsStore_SetUserPermissions(db database.DB) func(*testing.T) {
+	if permissions == nil {
+		permissions = []authz.Permission{}
+	}
+
+	if diff := cmp.Diff(expectedPermissions, permissions, cmpopts.IgnoreFields(authz.Permission{}, "CreatedAt", "UpdatedAt")); diff != "" {
+		t.Fatalf("Expected permissions: %v do not match actual permissions: %v; diff %v", expectedPermissions, permissions, diff)
+	}
+}
+
+func setupPermsRelatedEntities(t *testing.T, s *permsStore, permissions []authz.Permission) error {
+	t.Helper()
+	if permissions == nil || len(permissions) == 0 {
+		return errors.New("no permissions to setup foreign keys for")
+	}
+
+	users := make(map[int32]*sqlf.Query, len(permissions))
+	externalAccounts := make(map[int32]*sqlf.Query, len(permissions))
+	repos := make(map[int32]*sqlf.Query, len(permissions))
+	for _, p := range permissions {
+		users[p.UserID] = sqlf.Sprintf("(%s::integer, %s::text)", p.UserID, fmt.Sprintf("user-%d", p.UserID))
+		externalAccounts[p.ExternalAccountID] = sqlf.Sprintf("(%s::integer, %s::integer, %s::text, %s::text, %s::text, %s::text)", p.ExternalAccountID, p.UserID, "service_type", "service_id", fmt.Sprintf("account_id_%d", p.ExternalAccountID), "client_id")
+		repos[p.RepoID] = sqlf.Sprintf("(%s::integer, %s::text)", p.RepoID, fmt.Sprintf("repo-%d", p.RepoID))
+	}
+
+	usersQuery := sqlf.Sprintf(`INSERT INTO users(id, username) VALUES %s ON CONFLICT (id) DO NOTHING`, sqlf.Join(maps.Values(users), ","))
+	if err := s.execute(context.Background(), usersQuery); err != nil {
+		return err
+	}
+	externalAccountsQuery := sqlf.Sprintf(`INSERT INTO user_external_accounts(id, user_id, service_type, service_id, account_id, client_id) VALUES %s ON CONFLICT(id) DO NOTHING`, sqlf.Join(maps.Values(externalAccounts), ","))
+	if err := s.execute(context.Background(), externalAccountsQuery); err != nil {
+		return err
+	}
+	reposQuery := sqlf.Sprintf(`INSERT INTO repo(id, name) VALUES %s ON CONFLICT(id) DO NOTHING`, sqlf.Join(maps.Values(repos), ","))
+	if err := s.execute(context.Background(), reposQuery); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func testPermsStore_SetUserRepoPermissions(db database.DB) func(*testing.T) {
+	source := "test"
+
+	tests := []struct {
+		name                string
+		origPermissions     []authz.Permission
+		permissions         []authz.Permission
+		expectedPermissions []authz.Permission
+		entity              authz.PermissionEntity
+	}{
+		{
+			name:                "empty",
+			permissions:         []authz.Permission{},
+			expectedPermissions: []authz.Permission{},
+			entity: authz.PermissionEntity{
+				UserID:            1,
+				ExternalAccountID: 1,
+			},
+		},
+		{
+			name: "add",
+			permissions: []authz.Permission{
+				{
+					UserID:            1,
+					ExternalAccountID: 1,
+					RepoID:            1,
+				},
+				{
+					UserID:            1,
+					ExternalAccountID: 1,
+					RepoID:            2,
+				},
+				{
+					UserID:            1,
+					ExternalAccountID: 1,
+					RepoID:            3,
+				},
+			},
```

Comment:
```suggestion
		{
			name: "add",
			permissions: []authz.Permission{
				{UserID: 1, ExternalAccountID: 1, RepoID: 1},
				{UserID: 1, ExternalAccountID: 1, RepoID: 2},
				{UserID: 1, ExternalAccountID: 1, RepoID: 3},
			},
			expectedPermissions: []authz.Permission{
				{UserID: 1, ExternalAccountID: 1, RepoID: 1},
				{UserID: 1, ExternalAccountID: 1, RepoID: 2},
				{UserID: 1, ExternalAccountID: 1, RepoID: 3},
			},
			entity: authz.PermissionEntity{UserID: 1, ExternalAccountID: 1},
		},
```

Nitpick: I do think these still fit on one line and are actually more readable on one line :)

Reply:
Agreed. It was my editor that separated it into multiple lines. Will try to squeeze it to one-liners now.